File: Data Structure/Linked List/Linked List Cycle II/detectCycle_s_f.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        ListNode* slow=head;
        ListNode* fast=head;
        while(1)
        {            
            if(fast&&fast->next)
            {fast=fast->next->next;}
            else{return NULL;}
            slow=slow->next;
            if(fast==slow)
            {
                break;
            }       
        }
        slow=head;
        while(slow!=fast)
        {
            slow=slow->next;
            fast=fast->next;          
        }
        return fast;
    }
};

File: Data Structure/Linked List/Linked List Cycle II/detectCycle_s_f.py
class Solution(object):
    def detectCycle(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        slow=head;
        fast=head;
        try:
            while 1:
                slow=slow.next
                fast=fast.next.next
                if slow==fast:
                    break
        except:return None
        
        slow=head
        while fast!=slow:
            slow=slow.next
            fast=fast.next
        
        return slow
            
